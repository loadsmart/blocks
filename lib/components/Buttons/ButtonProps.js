"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ButtonDisplayState;
(function (ButtonDisplayState) {
    ButtonDisplayState[ButtonDisplayState["Normal"] = 0] = "Normal";
    ButtonDisplayState[ButtonDisplayState["Disabled"] = 1] = "Disabled";
    ButtonDisplayState[ButtonDisplayState["Loading"] = 2] = "Loading";
})(ButtonDisplayState = exports.ButtonDisplayState || (exports.ButtonDisplayState = {}));
var ButtonHeight;
(function (ButtonHeight) {
    ButtonHeight[ButtonHeight["Regular"] = 50] = "Regular";
    ButtonHeight[ButtonHeight["Short"] = 42] = "Short";
})(ButtonHeight = exports.ButtonHeight || (exports.ButtonHeight = {}));
