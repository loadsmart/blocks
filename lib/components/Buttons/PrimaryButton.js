"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var react_native_1 = require("react-native");
var res_1 = require("../../res/");
var ButtonProps_1 = require("./ButtonProps");
var PrimaryButton = /** @class */ (function (_super) {
    __extends(PrimaryButton, _super);
    function PrimaryButton() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.renderNormalState = function () {
            var _a = _this.props, icon = _a.icon, iconStyle = _a.iconStyle, title = _a.title;
            return (react_1.default.createElement(react_1.default.Fragment, null,
                icon && react_1.default.createElement(react_native_1.Image, { source: icon, style: [styles.icon, iconStyle] }),
                react_1.default.createElement(react_native_1.Text, { style: styles.text }, title.toUpperCase())));
        };
        _this.renderLoadingState = function () {
            return react_1.default.createElement(react_native_1.ActivityIndicator, { color: res_1.Colors.White, style: styles.activityIndicator });
        };
        _this.onPress = function () {
            var _a = _this.props, displayState = _a.displayState, onPress = _a.onPress;
            if (displayState === ButtonProps_1.ButtonDisplayState.Normal && onPress) {
                onPress();
            }
        };
        return _this;
    }
    PrimaryButton.prototype.render = function () {
        return (react_1.default.createElement(react_native_1.TouchableOpacity, { onPress: this.onPress.bind(this) },
            react_1.default.createElement(react_native_1.View, { style: [styles.wrapper, this.props.style] },
                this.props.displayState !== ButtonProps_1.ButtonDisplayState.Loading && this.renderNormalState(),
                this.props.displayState === ButtonProps_1.ButtonDisplayState.Loading && this.renderLoadingState())));
    };
    PrimaryButton.defaultProps = {
        displayState: ButtonProps_1.ButtonDisplayState.Normal,
    };
    return PrimaryButton;
}(react_1.PureComponent));
exports.default = PrimaryButton;
var styles = react_native_1.StyleSheet.create({
    wrapper: {
        flexDirection: 'row',
        borderRadius: 28,
        height: 50,
        paddingHorizontal: 20,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: res_1.Colors.AlgaeGreen,
    },
    icon: {
        marginRight: 8,
    },
    text: {
        color: 'white',
        fontFamily: res_1.Fonts.SharpSansExtrabold,
        fontSize: 15,
        alignSelf: 'center',
    },
    activityIndicator: {
        alignSelf: 'center',
        position: 'absolute',
    },
});
