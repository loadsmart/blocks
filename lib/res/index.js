"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var colors_1 = __importDefault(require("./colors"));
exports.Colors = colors_1.default;
var fonts_1 = __importDefault(require("./fonts"));
exports.Fonts = fonts_1.default;
var images_1 = __importDefault(require("./images"));
exports.Images = images_1.default;
