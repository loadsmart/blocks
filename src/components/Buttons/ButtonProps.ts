import { ImageSourcePropType, ImageStyle, StyleProp, ViewProps } from 'react-native'

export enum ButtonDisplayState {
  Normal,
  Disabled,
  Loading,
}

export enum ButtonHeight {
  Regular = 50,
  Short = 42,
}

export interface ButtonProps extends ViewProps {
  icon?: ImageSourcePropType
  iconStyle?: StyleProp<ImageStyle>
  title: string
  buttonHeight?: ButtonHeight
  displayState?: ButtonDisplayState
  onPress?: () => void
}
