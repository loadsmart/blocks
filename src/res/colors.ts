const Colors = {
  AlgaeGreen: '#14d64d',
  BrownGreyThree: '#9b9b9b',
  Charcoal: '#3c3d3a',
  CoralPink: '#fd576a',
  DarkGrey: '#666666',
  OrangePeel: '#ff9800',
  LightGray: '#d4d4d4',
  Tuna: '#484a4c',
  VeryLightGray: '#eee',
  White: '#fff',
  Ghost: '#c4ccd6',
  Transparent: '#00000000',
}

export default Colors
