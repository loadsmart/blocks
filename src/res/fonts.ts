const Fonts = {
  SharpSansBold: 'SharpSans-Bold',
  SharpSansItalic: 'SharpSans-BoldItalic',
  SharpSansBook: 'SharpSans-Book',
  SharpSansBookItalic: 'SharpSans-BookItalic',
  SharpSansExtrabold: 'SharpSans-Extrabold',
  SharpSansExtraboldItalic: 'SharpSans-ExtraboldItalic',
  SharpSansLight: 'SharpSans-Light',
  SharpSansLightItalic: 'SharpSans-LightItalic',
  SharpSansMedium: 'SharpSans-Medium',
  SharpSansMediumItalic: 'SharpSans-MediumItalic',
  SharpSansSemibold: 'SharpSans-Semibold',
  SharpSansSemiboldItalic: 'SharpSans-SemiboldItalic',
  SharpSansThin: 'SharpSans-Thin',
  SharpSansThinItalic: 'SharpSans-ThinItalic',
}

export default Fonts
