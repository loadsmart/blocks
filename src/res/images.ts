const Images = {
  Chevron: require('../../assets/images/chevron.png'),
  PlusSign: require(`../../assets/images/plus-sign.png`),
  SmallCheck: require(`../../assets/images/small-check.png`),
  Warning: require(`../../assets/images/warning.png`),
  PasswordShow: require('../../assets/images/password-show.png'),
  PasswordHide: require('../../assets/images/password-hide.png'),
}

export default Images
