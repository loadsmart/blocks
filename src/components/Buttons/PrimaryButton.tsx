import React, { PureComponent } from 'react'
import { ActivityIndicator, Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { Colors, Fonts } from '../../res/'
import { Theme, ThemeContext } from '../Contexts'
import { ButtonDisplayState, ButtonProps } from './ButtonProps'

export default class PrimaryButton extends PureComponent<ButtonProps> {
  public static defaultProps = {
    displayState: ButtonDisplayState.Normal,
  }

  public render() {
    const { displayState, testID } = this.props
    return (
      <ThemeContext.Consumer>
        {theme => {
          const styles = themedStyles(theme, this.props.style)
          return (
            <TouchableOpacity onPress={this.onPress.bind(this)} {...{ testID }}>
              <View style={[styles.wrapper, this.props.style]}>
                {displayState !== ButtonDisplayState.Loading && this.renderNormalState(styles)}
                {displayState === ButtonDisplayState.Loading && this.renderLoadingState(styles)}
              </View>
            </TouchableOpacity>
          )
        }}
      </ThemeContext.Consumer>
    )
  }

  private renderNormalState = (styles: any) => {
    const { icon, iconStyle, title } = this.props
    return (
      <>
        {icon && <Image source={icon} style={[styles.icon, iconStyle]} />}
        <Text style={styles.text}>{title.toUpperCase()}</Text>
      </>
    )
  }

  private renderLoadingState = (styles: any) => {
    return (
      <ActivityIndicator color={styles.activityIndicator.color} style={styles.activityIndicator} />
    )
  }

  private onPress = () => {
    const { displayState, onPress } = this.props
    if (displayState === ButtonDisplayState.Normal && onPress) {
      onPress()
    }
  }
}

const themedStyles = (theme: Partial<Theme>, style: any) => {
  return StyleSheet.create({
    wrapper: {
      flexDirection: 'row',
      borderRadius: 28,
      height: 50,
      paddingHorizontal: 20,
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: theme.primaryColor,
    },
    icon: {
      marginRight: 8,
    },
    text: {
      alignSelf: 'center',
      fontFamily: Fonts.SharpSansExtrabold,
      color: style && style.color ? style.color : Colors.White,
      fontSize: style && style.fontSize ? style.fontSize : 15,
      fontWeight: style && style.fontWeight ? style.fontWeight : 'normal',
    },
    activityIndicator: {
      alignSelf: 'center',
      position: 'absolute',
      color: style && style.color ? style.color : Colors.White,
    },
  })
}
