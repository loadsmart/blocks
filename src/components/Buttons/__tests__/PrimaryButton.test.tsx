import React from 'react'
import { ActivityIndicator, Text, TouchableOpacity, Image } from 'react-native'
import renderer from 'react-test-renderer'
import PrimaryButton from '../PrimaryButton'
import { ButtonDisplayState, ButtonProps } from '../ButtonProps'
import { Images } from '../../../res'

describe('PrimaryButton', () => {
  describe('when state is normal', () => {
    let testRenderer: renderer.ReactTestRenderer
    const onPressCallback = jest.fn()

    beforeEach(() => {
      const props: ButtonProps = {
        icon: Images.Warning,
        title: 'accept',
        displayState: ButtonDisplayState.Normal,
        onPress: onPressCallback,
      }
      testRenderer = renderer.create(<PrimaryButton {...props} />)
    })

    it('renders an icon', () => {
      const icon = testRenderer.root.findByType(Image)
      expect(icon.props.source).toBe(Images.Warning)
    })

    it('renders title', () => {
      const title = testRenderer.root.findByType(Text)
      expect(title.props.children).toBe('ACCEPT')
    })

    it('does not render spinner', () => {
      expect(testRenderer.root.findAllByType(ActivityIndicator)).toHaveLength(0)
    })

    it('will execute onPress callback', () => {
      const touchable = testRenderer.root.findByType(TouchableOpacity)
      touchable.props.onPress()
      expect(onPressCallback).toHaveBeenCalledTimes(1)
    })

    it('has a valid snapshot', () => {
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when state is loading', () => {
    let testRenderer: renderer.ReactTestRenderer
    const onPressCallback = jest.fn()

    beforeEach(() => {
      const props: ButtonProps = {
        title: 'accept',
        displayState: ButtonDisplayState.Loading,
        onPress: onPressCallback,
      }
      testRenderer = renderer.create(<PrimaryButton {...props} />)
    })

    it('does not render icon', () => {
      expect(testRenderer.root.findAllByType(Image)).toHaveLength(0)
    })

    it('does not render title', () => {
      expect(testRenderer.root.findAllByType(Text)).toHaveLength(0)
    })

    it('renders spinner', () => {
      expect(testRenderer.root.findAllByType(ActivityIndicator)).toHaveLength(1)
    })

    it("won't execute onPress callback", () => {
      const touchable = testRenderer.root.findByType(TouchableOpacity)
      touchable.props.onPress()
      expect(onPressCallback).not.toHaveBeenCalled()
    })

    it('has a valid snapshot', () => {
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when state is disabled', () => {
    let testRenderer: renderer.ReactTestRenderer
    const onPressCallback = jest.fn()

    beforeEach(() => {
      const props: ButtonProps = {
        title: 'accept',
        displayState: ButtonDisplayState.Disabled,
        onPress: onPressCallback,
      }
      testRenderer = renderer.create(<PrimaryButton {...props} />)
    })

    it('renders title', () => {
      const title = testRenderer.root.findByType(Text)
      expect(title.props.children).toBe('ACCEPT')
    })

    it('does not render spinner', () => {
      expect(testRenderer.root.findAllByType(ActivityIndicator)).toHaveLength(0)
    })

    it("won't execute onPress callback", () => {
      const touchable = testRenderer.root.findByType(TouchableOpacity)
      touchable.props.onPress()
      expect(onPressCallback).not.toHaveBeenCalled()
    })

    it('has a valid snapshot', () => {
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })
})
