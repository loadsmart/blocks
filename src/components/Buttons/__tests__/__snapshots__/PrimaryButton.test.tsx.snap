// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryButton when state is disabled has a valid snapshot 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#14d64d",
          "borderRadius": 28,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontFamily": "SharpSans-Extrabold",
          "fontSize": 15,
        }
      }
    >
      ACCEPT
    </Text>
  </View>
</View>
`;

exports[`PrimaryButton when state is loading has a valid snapshot 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#14d64d",
          "borderRadius": 28,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#fff"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "alignSelf": "center",
          "position": "absolute",
        }
      }
    />
  </View>
</View>
`;

exports[`PrimaryButton when state is normal has a valid snapshot 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#14d64d",
          "borderRadius": 28,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../assets/images/warning.png",
        }
      }
      style={
        Array [
          Object {
            "marginRight": 8,
          },
          undefined,
        ]
      }
    />
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontFamily": "SharpSans-Extrabold",
          "fontSize": 15,
        }
      }
    >
      ACCEPT
    </Text>
  </View>
</View>
`;
